Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,459 @@
+                self.terminate()
+                return
+            }
+        }
+    }
+
+    /// Yield new elements to the underlying `NIOAsyncSequenceProducer`.
+    func yield(_ element: Element) {
+        let action = self.stateMachine.withLockedValue { $0.yield(element) }
+        switch action {
+        case .produceMore:
+            self.produceMore()
+        case .stopProducing:
+            self.stopProducing()
+        case .terminate:
+            self.terminate()
+        case .none:
+            break
+        }
+    }
+
+    /// Determines the next action to be taken in the poll loop based on the current state.
+    ///
+    /// - Returns: The next action for the poll loop.
+    func nextPollLoopAction() -> KafkaPollingSystem.StateMachine.PollLoopAction {
+        return self.stateMachine.withLockedValue { $0.nextPollLoopAction() }
+    }
+
+    /// Stop producing new elements to the
+    /// `source` ``NIOAsyncSequenceProducer``.
+    private func stopProducing() {
+        self.stateMachine.withLockedValue { $0.stopProducing() }
+    }
+
+    /// Kill the ``KafkaPollingSystem``'s poll loop and free its resources.
+    func terminate(_ error: Error? = nil) {
+        let action = self.stateMachine.withLockedValue { $0.terminate(error) }
+        self.handleTerminateAction(action)
+    }
+
+    /// Invokes the desired action after ``KafkaPollingSystem/StateMachine/terminate()``
+    /// has been invoked.
+    private func handleTerminateAction(_ action: StateMachine.TerminateAction?) {
+        switch action {
+        case .finishSequenceSource(let source):
+            source?.finish()
+        case .finishSequenceSourceAndResume(let source, let continuation):
+            source?.finish()
+            continuation?.resume()
+        case .finishSequenceSourceAndResumeWithError(let source, let continuation, let error):
+            source?.finish()
+            continuation?.resume(throwing: error)
+        case .none:
+            break
+        }
+    }
+}
+
+extension KafkaPollingSystem: NIOAsyncSequenceProducerDelegate {
+    func produceMore() {
+        let action = self.stateMachine.withLockedValue { $0.produceMore() }
+        switch action {
+        case .resume(let continuation):
+            continuation?.resume()
+        case .none:
+            break
+        }
+    }
+
+    func didTerminate() {
+        self.terminate()
+    }
+}
+
+extension KafkaPollingSystem {
+    /// The state machine used by the ``KafkaPollingSystem``.
+    struct StateMachine {
+        /// The possible states of the state machine.
+        enum State {
+            /// The state machine has been initialized with init() but is not yet Initialized
+            /// using `func initialize()` (required).
+            case uninitialized
+            /// Initialized state (idle).
+            case idle(
+                source: Producer.Source?,
+                pollClosure: (() -> Void)?,
```

Comment:
Why are the `source` and the `pollClosure` optional here?

Reply:
You're right, this was from a previous implementation!